### PR TITLE
convert samples page to use catalog extensions

### PR DESCRIPTION
**Pull request by christianvogt**
Converts the samples page to use the same catalog extensions as the dev-catalog. In doing so we gain access to keyword search and sorting. This also prepares the samples page for the addition of devfile samples.

Updated catalog extensions to receive an additional `catalogId` property in order to allow for contributing catalog providers to different catalogs.

Added capability to catalog components to forcefully hide the side bar as well as execute the CTA immediately on clicking the catalog tile instead of opening up a details panel. These options are configured at render time of each individual catalog.

**Screen shots / Gifs for design review**: 

fyi @openshift/team-devconsole-ux
image

**Browser conformance**: 

- [ ] Chrome
- [ ] Firefox
- [ ] Safari
- [x] Edge

/assign @rohitkrai03

### CODE REVIEW

**Comment by beaumorley** [comment]
lgtm. nice update.

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/catalog.ts
Comment (rohitkrai03) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/catalog.ts:23]: Should this also be?
```suggestion
    catalogId: string | string[];
```
This way one provider can provide to multiple catalogs. For example we have `dev-catalog` and `topology-quick-add-catalog` getting almost same catalog items but quick starts is only supposed to be in topology quick add. There might be other items exclusive for quick add in future.

File: frontend/packages/topology/src/components/quick-search/QuickSearch.tsx
Comment (rohitkrai03) [inline frontend/packages/topology/src/components/quick-search/QuickSearch.tsx:69]: This should probably use a different catalogId because the items in this catalog might be different than the dev catalog.

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/catalog.ts
Comment (christianvogt) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/catalog.ts:23]: Actually I'll make it just an array to avoid the different types.

Comment (christianvogt) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/catalog.ts]: nm... i made your suggestion

File: frontend/packages/dev-console/package.json
Comment (christianvogt) [reply frontend/packages/dev-console/package.json]: relocated useBuilderImageSamples and used the same exposed module

**Comment by christianvogt** [comment]
@rohitkrai03 i've made the suggested changes

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/catalog.ts
Comment (rohitkrai03) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/catalog.ts]: I actually prefer `string[]` myself, makes it simpler for the extension consumer. My suggestion was based on our earlier extensions where we use `string | string[]` for `insertBefore` and `insertAfter`.

**Review by rohitkrai03 — APPROVED** [review]
/lgtm

**Comment by spadgett** [comment]
/hold

@christianvogt I believe this is causing batch merges to fail. Possibly it has build errors when run with another PR in the merge queue. Can you take a look?

https://prow.ci.openshift.org/?repo=openshift%2Fconsole&job=pull-ci-openshift-console-master-e2e-gcp-console

```
ERROR in /go/src/github.com/openshift/console/frontend/packages/dev-console/src/components/catalog/providers/useBuilderImageSamples.ts
ERROR in /go/src/github.com/openshift/console/frontend/packages/dev-console/src/components/catalog/providers/useBuilderImageSamples.ts(3,10):
TS2305: Module '"../../../../../console-dynamic-plugin-sdk/src"' has no exported member 'CatalogExtensionHook'.
```

https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/batch/pull-ci-openshift-console-master-e2e-gcp-console/1377503103363846144

**Comment by christianvogt** [comment]
@spadgett yes this PR will conflict with https://github.com/openshift/console/pull/8467

We can hold until this PR merges and I'll rebase mine then.

**Comment by christianvogt** [comment]
rebased now that https://github.com/openshift/console/pull/8467 has merged

**Pull request merged**
